### PR TITLE
Speed up tidy

### PR DESCRIPTION
**Pull request by the8472**
This can be reviewed commit by commit since they contain separate optimizations.

```
# master
$ taskset -c 0-5 hyperfine './x test tidy'
Benchmark #1: ./x test tidy
  Time (mean ± σ):      4.857 s ±  0.064 s    [User: 12.967 s, System: 2.014 s]
  Range (min … max):    4.779 s …  4.997 s    10 runs

# PR
$ taskset -c 0-5 hyperfine './x test tidy'
Benchmark #1: ./x test tidy
  Time (mean ± σ):      3.672 s ±  0.035 s    [User: 10.524 s, System: 2.029 s]
  Range (min … max):    3.610 s …  3.725 s    10 runs
```

### CODE REVIEW

**Review by jyn514 — APPROVED** [review]
Thanks for working on this :) I don't mind merging as-is, but the concurrency code here is getting pretty hairy, I wonder if it makes sense to switch to rayon.

File: src/tools/tidy/src/style.rs [hunk 1]
```diff
@@ -281,7 +282,27 @@ pub fn check(path: &Path, bad: &mut bool) {
...
+                continue;
+            } else {
+                trailing_new_lines = 0;
+            }
+
+            let trimmed = line.trim();
+
+            if !trimmed.starts_with("//") {
+                lines += 1;
+            }
```
[the remark this answers is not in the document]
Comment (the8472) [reply src/tools/tidy/src/style.rs:304]: So we can `continue` earlier and skip attempting all the string processing stuff below on empty lines.

File: src/tools/tidy/src/main.rs [hunk 2]
```diff
@@ -35,15 +35,26 @@ fn main() {
 
     let bad = std::sync::Arc::new(AtomicBool::new(false));
 
+    let drain_handles = |handles: &mut VecDeque<ScopedJoinHandle<'_, ()>>| {
+        // poll all threads for completion before awaiting the oldest one
+        for i in (0..handles.len()).rev() {
+            if handles[i].is_finished() {
+                handles.swap_remove_back(i).unwrap().join().unwrap();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (the8472) [reply src/tools/tidy/src/main.rs:44]: This is called every time in the `check` macro when a task is added to make room for the next task.

When I originally wrote this code (before `is_finished` existed) I think I wanted to add rayon but someone said it's better to keep the number of dependencies low so it's faster when one doesn't have tidy compiled.

[the remark this answers is not in the document]
Comment (jyn514) [reply src/tools/tidy/src/main.rs:44]: Hmm, the only place I see rayon mentioned in the original PR is https://github.com/rust-lang/rust/pull/81833#issuecomment-777855100 - do you remember why you made that change?

[the remark this answers is not in the document]
Comment (jyn514) [reply src/tools/tidy/src/main.rs:44]: ah, looks like https://github.com/rust-lang/rust/pull/81833#issuecomment-782756504. Ok, we don't need to revisit that here.

**Pull request merged**